File: cloudflare/control-plane/src/workspace.ts
import { DurableObject } from "cloudflare:workers";
import { authorityFailure, digest, Fault, Json, Principal, requireThat, stable, State, validate } from "./shared";
import { artifactManifests, artifactReadLocator, claimArtifactUpload, cleanupArtifactObjects, commitArtifactPublication, initializeArtifacts, prepareArtifactPublication, pruneArtifactReceipts } from "./artifacts";
import { checkpoint, recovery, registerProject, registerSource } from "./collector";
import { dropEmptyLegacyFactTables, legacyFactTables } from "./legacy-cleanup";
import { livingRead, livingWrite } from "./living";
import { deleteWorkspaceArtifactPrefix, initializeReplacement, markWorkspaceReplacement, previewWorkspaceReplacement, workspaceReplacement } from "./replacement";

const REPLACEMENT_MUTATIONS = new Set([
  "ct_project_register", "ct_collector_register_source", "ct_collector_publish_observation",
  "ct_collector_publish_artifacts", "ct_collector_heartbeat", "ct_collector_publish_living_observation",
  "ct_internal_artifact_claim",
]);

/** A workspace is the authorization, transaction and revision boundary. */
export class Workspace extends DurableObject<Env> {
  private state: State;
  constructor(ctx: DurableObjectState, env: Env) {
    super(ctx, env);
    this.state = new State(ctx.storage.sql);
    this.initialize();
  }

  private initialize() {
    this.ctx.storage.transactionSync(() => {
      initializeArtifacts(this.state);
      initializeReplacement(this.state);
      const target = this.env.CT_LEGACY_FACT_CLEANUP_WORKSPACE_ID;
      if (target && this.ctx.id.toString() === this.env.WORKSPACES.idFromName(target).toString()) {
        dropEmptyLegacyFactTables(this.state);
      }
    });
  }

  async invoke(method: string, envelopeJson: string, principalJson: string): Promise<string> {
    const envelope = JSON.parse(envelopeJson);
    const principal = JSON.parse(principalJson);
    return JSON.stringify(await this.rpc(method, envelope, principal));
  }

  async rpc(method: string, envelope: Json, principal: Principal): Promise<{ status: number; body: Json }> {
    try {
      const request = envelope.request;
      requireThat(request.workspace_id === principal.workspace_id, "workspace_denied", 403);
      if (method === "ct_workspace_replace") {
        requireThat(principal.roles.includes("owner"), "capability_required", 403);
        requireThat(this.env.CT_REPLACEMENT_WORKSPACE_ID
          && request.workspace_id === this.env.CT_REPLACEMENT_WORKSPACE_ID,
        "workspace_replacement_target_denied", 403);
        requireThat(this.env.CT_REPLACEMENT_EXPORT_SHA256
          && request.expected_export_sha256 === this.env.CT_REPLACEMENT_EXPORT_SHA256,
        "workspace_replacement_export_denied", 403);
        requireThat(request.confirmation
          === `${request.mode}:${request.workspace_id}:${request.expected_export_sha256}`,
        "workspace_replacement_confirmation_required", 403);
        const expectedObject = this.env.WORKSPACES.idFromName(request.workspace_id).toString();
        requireThat(this.ctx.id.toString() === expectedObject, "workspace_replacement_target_denied", 403);
        if (request.mode === "preview") {
          return { status: 200, body: await previewWorkspaceReplacement(
            this.state, this.env, request.workspace_id, request.expected_export_sha256,
          ) };
        }
        requireThat(request.mode === "execute", "workspace_replacement_mode_invalid");
        return this.ctx.blockConcurrencyWhile(async () => {
          const prior = workspaceReplacement(this.state);
          const same = prior != null && prior.workspace_id === request.workspace_id
            && prior.export_sha256 === request.expected_export_sha256;
          if (prior && same && prior.status === "complete") {
            return { status: 200, body: {
              workspace_id: request.workspace_id, sql_reset: false,
              prefix: `workspaces/${request.workspace_id}/artifacts/`,
              deleted: 0, complete: true, already_complete: true,
            } };
          }
          const resetSql = !same;
          if (!same) {
            await this.ctx.storage.deleteAll();
            this.state = new State(this.ctx.storage.sql);
            this.initialize();
            markWorkspaceReplacement(this.state, request.workspace_id,
              request.expected_export_sha256, "incomplete");
          }
          const deletion = await deleteWorkspaceArtifactPrefix(this.env, request.workspace_id);
          markWorkspaceReplacement(this.state, request.workspace_id,
            request.expected_export_sha256, deletion.complete ? "complete" : "incomplete");
          return { status: 200, body: {
            workspace_id: request.workspace_id, sql_reset: resetSql,
            ...deletion, already_complete: false,
          } };
        });
      }
      const replacement = workspaceReplacement(this.state);
      requireThat(replacement?.status !== "incomplete" || !REPLACEMENT_MUTATIONS.has(method),
        "workspace_replacement_incomplete", 409);
      if (method === "ct_internal_artifact_claim") {
        requireThat(principal.roles.includes("collect") || principal.roles.includes("owner"), "capability_required", 403);
        requireThat(["facts", "summary"].includes(request.kind)
          && typeof request.sha256 === "string"
          && /^[0-9a-f]{64}$/.test(request.sha256), "invalid_artifact_claim");
        this.ctx.storage.transactionSync(() =>
          claimArtifactUpload(this.state, request.kind, request.sha256));
        return { status: 200, body: {} };
      }
      // Compute identity before schema defaults normalize the request.
      const identity = await digest(stable(request));
      validate(method, request);
      if (method === "ct_artifact_read") {
        return { status: 200, body: artifactReadLocator(this.state, request) };
      }
      if (method === "ct_collector_publish_artifacts") {
        const key = envelope.idempotency_key
          ? stable([principal.agent_id, method, envelope.idempotency_key])
          : null;
        const prior = key ? this.state.get("receipt", key) : undefined;
        if (prior) {
          requireThat(prior.identity === identity, "idempotency_conflict", 409);
          return { status: 200, body: prior.result };
        }
        const plan = await prepareArtifactPublication(this.state, this.env, request);
        const body = this.ctx.storage.transactionSync(() => {
          const concurrent = key ? this.state.get("receipt", key) : undefined;
          if (concurrent) {
            requireThat(concurrent.identity === identity, "idempotency_conflict", 409);
            return concurrent.result;
          }
          const result = commitArtifactPublication(this.state, request, plan);
          if (key) this.state.put("receipt", key, { identity, result }, this.state.head());
          pruneArtifactReceipts(this.state);
          return result;
        });
        this.ctx.waitUntil(this.ctx.blockConcurrencyWhile(() =>
          cleanupArtifactObjects(this.state, this.env, request.workspace_id)
            .catch(() => console.error(JSON.stringify({ event: "artifact_cleanup_deferred" })))));
        return { status: 200, body };
      }
      if (method === "ct_collector_publish_observation") {
        requireThat(request.payload.source_checkpoint?.segments?.every((offset: unknown) => Number.isSafeInteger(offset) && Number(offset) > 0), "invalid_checkpoint_offsets");
        requireThat(await digest(stable(request.payload)) === request.content_sha256 && request.event_id === `checkpoint:${request.content_sha256}`, "checkpoint_digest_mismatch");
      }
      const body = this.ctx.storage.transactionSync(() => {
        const key = envelope.idempotency_key ? stable([principal.agent_id, method, envelope.idempotency_key]) : null;
        const prior = key ? this.state.get("receipt", key) : undefined;
        if (prior) {
          requireThat(prior.identity === identity, "idempotency_conflict", 409);
          return prior.result;
        }
        const result = this.dispatch(method, request, principal);
        if (key) this.state.put("receipt", key, { identity, result }, this.state.head());
        return result;
      });
      return { status: 200, body };
    } catch (error) {
      const failure = authorityFailure(error, "workspace");
      return { status: failure.status, body: { error: { code: failure.code } } };
    }
  }

  private dispatch(method: string, request: Json, principal: Principal): Json {
    switch (method) {
      case "ct_project_register": return registerProject(this.state, request);
      case "ct_collector_register_source": return registerSource(this.state, request);
      case "ct_collector_recover": return recovery(this.state, request);
      case "ct_collector_publish_observation": validate("checkpoint", request.payload); return checkpoint(this.state, request);
      case "ct_collector_heartbeat": case "ct_collector_publish_living_observation": return livingWrite(this.state, method, request);
      case "ct_remote_living": return livingRead(this.state, request);
      case "ct_workspace_snapshot": return { workspace_id: request.workspace_id, snapshot_sequence: this.state.pin(request.snapshot_sequence) };
      case "ct_legacy_fact_cleanup_status": return {
        workspace_id: request.workspace_id, snapshot_sequence: this.state.head(),
        ...legacyFactTables(this.state),
      };
      case "ct_artifact_manifest": return artifactManifests(this.state, request);
      case "ct_project_inventory_snapshot": {
        const sequence = this.state.pin(request.snapshot_sequence);
        const artifactRows = this.state.sql.exec<{ project_id: string; manifest: string }>(`SELECT project_id,manifest
          FROM artifact_manifests current WHERE workspace_sequence=(
            SELECT MAX(candidate.workspace_sequence) FROM artifact_manifests candidate
            WHERE candidate.project_id=current.project_id AND candidate.workspace_sequence<=?)`, sequence).toArray();
        const expectedArtifactProjects = this.state.all("artifact_project_publisher", sequence)
          .filter(row => row.first_sequence <= sequence);
        if (artifactRows.length !== expectedArtifactProjects.length) {
          throw new Fault(410, "artifact_snapshot_expired");
        }
        const artifactGraphs = artifactRows.flatMap(row => JSON.parse(row.manifest).graphs.map((graph: Json) => ({
            project_id: row.project_id, vendors: graph.vendors,
          })));
        const projects = this.state.all("project", sequence)
          .filter(row => !request.modified_since || Date.parse(row.modified_at) >= Date.parse(request.modified_since))
          .map((row): Json => ({
            ...row,
            vendors: [...new Set(artifactGraphs.filter(graph => graph.project_id === row.project_id).flatMap(graph => graph.vendors))].sort(),
          }))
          .filter(row => !request.agent_vendor || row.vendors.includes(request.agent_vendor))
          .sort((a, b) => a.display_name.localeCompare(b.display_name));
        return { workspace_id: request.workspace_id, snapshot_sequence: sequence, projects };
      }
      default: throw new Fault(404, "not_found");
    }
  }
}

File: cloudflare/control-plane/src/workspace.py
"""Workspace Durable Object: authorization, transactions, and revision boundary."""

from __future__ import annotations

import asyncio
import json
import re
import sys
from typing import Any

from artifact_manifest import expand_manifest
from artifacts import (
    artifact_manifests,
    artifact_read_locator,
    artifact_readiness,
    claim_artifact_upload,
    cleanup_artifact_objects,
    commit_artifact_publication,
    complete_artifact_upload,
    initialize_artifacts,
    prepare_artifact_publication,
    prune_artifact_receipts,
)
from collector import checkpoint, recovery, register_project, register_source
from living import living_read, living_write
from prepared_api import api_locator, initialize_api
from shared import (
    Fault,
    State,
    authority_failure,
    digest,
    js_get,
    parse_timestamp,
    require_that,
    rows,
    stable,
    validate_contract,
)
from workers import DurableObject

Json = dict[str, Any]
Principal = dict[str, Any]


class Workspace(DurableObject):
    def __init__(self, ctx: Any, env: Any) -> None:
        super().__init__(ctx, env)
        self.state = State(ctx.storage.sql)
        self._invocation_lock = asyncio.Lock()
        self._initialize()

    def _initialize(self) -> None:
        def initialize() -> None:
            initialize_artifacts(self.state)
            initialize_api(self.state)

        self.ctx.storage.transactionSync(initialize)

    async def fetch(self, request: Any):
        from http_handler import HttpHandler

        return await HttpHandler(_HttpEnvironment(self)).fetch(request)

    async def invoke(self, method: str, envelope_json: str, principal_json: str) -> str:
        envelope = json.loads(envelope_json)
        principal = json.loads(principal_json)
        # A DO input gate does not preserve an operation across R2 awaits. Keep
        # source/claim/publication/cleanup calls ordered for this instance.
        async with self._invocation_lock:
            result = await self.rpc(method, envelope, principal)
        return json.dumps(result, ensure_ascii=False, separators=(",", ":"))

    async def rpc(self, method: str, envelope: Json, principal: Principal) -> Json:
        try:
            request = envelope["request"]
            require_that(
                request["workspace_id"] == principal["workspace_id"],
                "workspace_denied",
                403,
            )
            if method in {
                "ct_internal_artifact_claim",
                "ct_internal_artifact_complete",
            }:
                require_that(
                    "collect" in principal["roles"] or "owner" in principal["roles"],
                    "capability_required",
                    403,
                )
                require_that(
                    request.get("kind") in {"facts", "summary", "api"}
                    and isinstance(request.get("sha256"), str)
                    and re.fullmatch(r"[0-9a-f]{64}", request["sha256"]),
                    "invalid_artifact_claim",
                )
                if method == "ct_internal_artifact_claim":
                    token = self.ctx.storage.transactionSync(
                        lambda: claim_artifact_upload(
                            self.state, request["kind"], request["sha256"]
                        )
                    )
                    return {"status": 200, "body": {"token": token}}
                self.ctx.storage.transactionSync(
                    lambda: complete_artifact_upload(self.state, request)
                )
                return {"status": 200, "body": {}}
            if method == "ct_internal_api_locator":
                require_that(
                    "read" in principal["roles"] or "owner" in principal["roles"],
                    "capability_required",
                    403,
                )
                return {"status": 200, "body": api_locator(self.state, request)}

            # The idempotency identity covers raw JSON before defaults and typed
            # normalization alter its representation.
            identity = await digest(stable(request))
            if (
                method == "ct_collector_publish_artifacts"
                and request.get("schema_version") == "ct.artifact-manifest.v3"
            ):
                request = validate_contract("compact_publication", request)
                request = expand_manifest(request)
            request = validate_contract(method, request)
            if method == "ct_collector_artifact_readiness":
                require_that(
                    "collect" in principal["roles"] or "owner" in principal["roles"],
                    "capability_required",
                    403,
                )
                return {"status": 200, "body": artifact_readiness(self.state, request)}
            if method == "ct_artifact_read":
                return {
                    "status": 200,
                    "body": artifact_read_locator(self.state, request),
                }
            if method == "ct_remote_living":
                return {
                    "status": 200,
                    "body": await living_read(
                        self.state, request, self.env.CT_CURSOR_KEY
                    ),
                }
            if method == "ct_collector_publish_artifacts":
                key = (
                    stable([principal["agent_id"], method, envelope["idempotency_key"]])
                    if envelope.get("idempotency_key")
                    else None
                )
                prior = self.state.get("receipt", key) if key else None
                if prior:
                    require_that(
                        prior["identity"] == identity, "idempotency_conflict", 409
                    )
                    return {"status": 200, "body": prior["result"]}
                plan = await prepare_artifact_publication(self.state, self.env, request)

                def commit() -> Json:
                    concurrent = self.state.get("receipt", key) if key else None
                    if concurrent:
                        require_that(
                            concurrent["identity"] == identity,
                            "idempotency_conflict",
                            409,
                        )
                        return concurrent["result"]
                    result = commit_artifact_publication(self.state, request, plan)
                    if key:
                        self.state.put(
                            "receipt",
                            key,
                            {"identity": identity, "result": result},
                            self.state.head(),
                        )
                    prune_artifact_receipts(self.state)
                    return result

                body = self.ctx.storage.transactionSync(commit)
                try:
                    await cleanup_artifact_objects(
                        self.state, self.env, request["workspace_id"]
                    )
                except Exception:  # noqa: BLE001 -- cleanup is explicitly best-effort
                    print(
                        json.dumps({"event": "artifact_cleanup_deferred"}),
                        file=sys.stderr,
                    )
                return {"status": 200, "body": body}
            if method == "ct_collector_publish_observation":
                source_checkpoint = request["payload"].get("source_checkpoint")
                segments = (
                    source_checkpoint.get("segments", []) if source_checkpoint else []
                )
                require_that(
                    all(
                        isinstance(offset, int)
                        and not isinstance(offset, bool)
                        and offset > 0
                        for offset in segments
                    ),
                    "invalid_checkpoint_offsets",
                )
                require_that(
                    await digest(stable(request["payload"]))
                    == request["content_sha256"]
                    and request["event_id"]
                    == f"checkpoint:{request['content_sha256']}",
                    "checkpoint_digest_mismatch",
                )

            def dispatch_transaction() -> Json:
                key = (
                    stable([principal["agent_id"], method, envelope["idempotency_key"]])
                    if envelope.get("idempotency_key")
                    else None
                )
                prior = self.state.get("receipt", key) if key else None
                if prior:
                    require_that(
                        prior["identity"] == identity, "idempotency_conflict", 409
                    )
                    return prior["result"]
                result = self.dispatch(method, request, principal)
                if key:
                    self.state.put(
                        "receipt",
                        key,
                        {"identity": identity, "result": result},
                        self.state.head(),
                    )
                return result

            body = self.ctx.storage.transactionSync(dispatch_transaction)
            return {"status": 200, "body": body}
        except Exception as error:  # noqa: BLE001 -- sanitize every platform boundary failure
            failure = authority_failure(error, "workspace")
            return {"status": failure.status, "body": {"error": {"code": failure.code}}}

    def dispatch(self, method: str, request: Json, principal: Principal) -> Json:
        if method == "ct_project_register":
            return register_project(self.state, request)
        if method == "ct_collector_register_source":
            return register_source(self.state, request)
        if method == "ct_collector_recover":
            return recovery(self.state, request)
        if method == "ct_collector_publish_observation":
            request["payload"] = validate_contract("checkpoint", request["payload"])
            return checkpoint(self.state, request)
        if method in {
            "ct_collector_heartbeat",
            "ct_collector_publish_living_observation",
        }:
            return living_write(self.state, method, request)
        if method == "ct_workspace_snapshot":
            return {
                "workspace_id": request["workspace_id"],
                "snapshot_sequence": self.state.pin(request.get("snapshot_sequence")),
            }
        if method == "ct_artifact_manifest":
            return artifact_manifests(self.state, request)
        if method == "ct_project_inventory_snapshot":
            sequence = self.state.pin(request.get("snapshot_sequence"))
            artifact_rows = rows(
                self.state.sql.exec(
                    """SELECT project_id,manifest FROM artifact_manifests current WHERE workspace_sequence=(
                    SELECT MAX(candidate.workspace_sequence) FROM artifact_manifests candidate
                    WHERE candidate.project_id=current.project_id AND candidate.workspace_sequence<=?)""",
                    sequence,
                )
            )
            expected = [
                row
                for row in self.state.all("artifact_project_publisher", sequence)
                if row["first_sequence"] <= sequence
            ]
            if len(artifact_rows) != len(expected):
                raise Fault(410, "artifact_snapshot_expired")
            artifact_graphs: list[Json] = []
            for row in artifact_rows:
                for graph in json.loads(js_get(row, "manifest"))["graphs"]:
                    artifact_graphs.append(
                        {
                            "project_id": js_get(row, "project_id"),
                            "vendors": graph["vendors"],
                        }
                    )
            modified_since = (
                parse_timestamp(request["modified_since"]).timestamp()
                if request.get("modified_since")
                else None
            )
            projects = []
            for row in self.state.all("project", sequence):
                if (
                    modified_since is not None
                    and parse_timestamp(row["modified_at"]).timestamp() < modified_since
                ):
                    continue
                vendors = sorted(
                    {
                        vendor
                        for graph in artifact_graphs
                        if graph["project_id"] == row["project_id"]
                        for vendor in graph["vendors"]
                    }
                )
                project = {**row, "vendors": vendors}
                if (
                    request.get("agent_vendor")
                    and request["agent_vendor"] not in vendors
                ):
                    continue
                projects.append(project)
            projects.sort(key=lambda item: item["display_name"])
            return {
                "workspace_id": request["workspace_id"],
                "snapshot_sequence": sequence,
                "projects": projects,
            }
        raise Fault(404, "not_found")


class _LocalWorkspaceNamespace:
    """Resolve HTTP operations to this object without a second RPC hop."""

    def __init__(self, workspace: Workspace) -> None:
        self.workspace = workspace

    def getByName(self, name: str) -> Workspace:
        expected = self.workspace.env.WORKSPACES.idFromName(name).toString()
        require_that(
            expected == self.workspace.ctx.id.toString(), "workspace_denied", 403
        )
        return self.workspace


class _HttpEnvironment:
    def __init__(self, workspace: Workspace) -> None:
        self._env = workspace.env
        self.WORKSPACES = _LocalWorkspaceNamespace(workspace)

    def __getattr__(self, name: str) -> Any:
        return getattr(self._env, name)
